sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala

```diff
@@ -200,6 +228,11 @@ case object SinglePartition extends Partitioning {
  * Represents a partitioning where rows are split up across partitions based on the hash
  * of `expressions`.  All rows where `expressions` evaluate to the same values are guaranteed to be
  * in the same partition.
+ *
+ * Since [[StatefulOpClusteredDistribution]] relies on this partitioning and Spark requires
+ * stateful operators to retain the same physical partitioning during the lifetime of the query
+ * (including restart), the result of evaluation on `partitionIdExpression` must be unchanged
+ * across Spark versions. Violation of this requirement may bring silent correctness issue.
```

Comment:
shall we enforce this assumption in unit test as well? e.g. in `StreamingJoinSuite`. It's great to highlight in comment here, but people always forget and the unit test will fail loudly when we introduce any invalid change.

Reply:
We have a test for verifying this, although it is not exhaustive.

https://github.com/apache/spark/blob/2e703ae9d210d26573b849a9a88c55667b24127d/sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala#L574-L605

If we want to be exhaustive, I can make a combination of repartitions which could have not triggered shuffle with hash partitioning against joining keys if stream-stream join uses ClusteredDistribution. It may not be exhaustive for future-proof indeed.

Instead, if we are pretty sure StateOpClusteredDistribution would work as expected, we can simply check the required child distribution of the physical plan of stream-stream join.

Reply:
oh actually I was referring to the assumption:

* `HashPartitioning.partitionIdExpression` has to be exactly `Pmod(new Murmur3Hash(expressions), Literal(numPartitions))`.

It would be just to add some logic to check `opA/opB.partitionIdExpression` for the `opA/opB` at Line 598/599. I can also do it later if it's not clear to you.

Reply:
We checked HashPartitioning and partitionExpression here - the remaining is partitionIdExpression, which is the implementation of HashPartitioning.

That said, it would be nice if we have a separate test against HashPartitioning if we don't have one. Could you please check and craft one if we don't have it?

Reply:
Sure I can add one later this week.